File: ts/stepA_mal.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var printer_1 = require("./printer");
var reader_1 = require("./reader");
var env_1 = require("./env");
var core_1 = require("./core");
var node_readline_1 = require("./node_readline");
var repl_env = new env_1.Env();
repl_env.set("+", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value + b.value
        }); }
    }
});
repl_env.set("-", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value - b.value
        }); }
    }
});
repl_env.set("*", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value * b.value
        }); }
    }
});
repl_env.set("/", {
    type: "function",
    value: {
        fn: function (a, b) { return ({
            type: "number",
            value: a.value / b.value
        }); }
    }
});
repl_env.set("eval", {
    type: "function",
    value: {
        fn: function (ast) { return EVAL(ast, repl_env); }
    }
});
Object.entries(core_1.ns).forEach(function (_a) {
    var key = _a[0], value = _a[1];
    return repl_env.set(key, value);
});
var eval_ast = function (ast, env) {
    if (ast.type === "symbol") {
        return env.get(ast.value);
    }
    if (ast.type === "list") {
        return {
            type: "list",
            value: ast.value.map(function (val) { return EVAL(val, env); })
        };
    }
    if (ast.type === "vector") {
        return {
            type: "vector",
            value: ast.value.map(function (val) { return EVAL(val, env); })
        };
    }
    if (ast.type === "map") {
        return {
            type: "map",
            value: Object.entries(ast.value).reduce(function (acc, _a) {
                var key = _a[0], val = _a[1];
                acc[key] = EVAL(val, env);
                return acc;
            }, {})
        };
    }
    return ast;
};
var isPair = function (ast) {
    return (ast.type === "vector" || ast.type === "list") && ast.value.length !== 0;
};
var quasiquote = function (ast) {
    if (!isPair(ast)) {
        return {
            type: "list",
            value: [
                {
                    type: "symbol",
                    value: "quote"
                },
                ast
            ]
        };
    }
    if (ast.value[0].value === "unquote") {
        return ast.value[1];
    }
    if (isPair(ast.value[0])) {
        if (ast.value[0].value[0].value === "splice-unquote") {
            return {
                type: "list",
                value: [
                    {
                        type: "symbol",
                        value: "concat"
                    },
                    ast.value[0].value[1],
                    quasiquote({ type: "list", value: ast.value.slice(1) })
                ]
            };
        }
    }
    return {
        type: "list",
        value: [
            {
                type: "symbol",
                value: "cons"
            },
            quasiquote(ast.value[0]),
            quasiquote({ type: "list", value: ast.value.slice(1) })
        ]
    };
};
var isMacroCall = function (ast, env) {
    if (isPair(ast)) {
        if (ast.value[0].type == "symbol") {
            try {
                var value = env.get(ast.value[0].value);
                return value.type === "function" && value.is_macro === true;
            }
            catch (e) {
                return false;
            }
        }
    }
    return false;
};
var macroexpand = function (ast, env) {
    var _a;
    while (isMacroCall(ast, env)) {
        if (ast.value[0].type !== "symbol") {
            throw new Error("Should be a symbol");
        }
        var new_ast = env.get(ast.value[0].value);
        if (new_ast.type !== "function" || !new_ast.value.params) {
            throw new Error("Should be a custom func");
        }
        ast = (_a = new_ast.value).fn.apply(_a, ast.value.slice(1));
    }
    return ast;
};
var EVAL = function (ast, env) {
    var _loop_1 = function () {
        var _a;
        ast = macroexpand(ast, env);
        if (ast.type !== "list") {
            return { value: eval_ast(ast, env) };
        }
        if (ast.value.length === 0) {
            return { value: ast };
        }
        if (ast.value[0].value === "try*") {
            try {
                return { value: EVAL(ast.value[1], env) };
            }
            catch (e) {
                if (ast.value.length < 3) {
                    throw e;
                }
                if (ast.value[2].type !== "list") {
                    throw new Error("should be a list");
                }
                var _b = ast.value[2].value, errorName = _b[1], content = _b[2];
                if (errorName.type !== "symbol") {
                    throw new Error("should be a symbol");
                }
                var new_env = new env_1.Env(env);
                if (e instanceof Error) {
                    e = {
                        type: "string",
                        value: e.message
                    };
                }
                new_env.set(errorName.value, e);
                return { value: EVAL(content, new_env) };
            }
        }
        if (ast.value[0].value === "quote") {
            return { value: ast.value[1] };
        }
        if (ast.value[0].value === "macroexpand") {
            return { value: macroexpand(ast.value[1], env) };
        }
        if (ast.value[0].value === "quasiquote") {
            ast = quasiquote(ast.value[1]);
            return "continue";
        }
        if (ast.value[0].value === "def!") {
            if (ast.value[1].type !== "symbol") {
                throw new Error("Should be a symbol");
            }
            var evaluated_1 = EVAL(ast.value[2], env);
            if (evaluated_1.type !== "error") {
                env.set(ast.value[1].value, evaluated_1);
            }
            return { value: evaluated_1 };
        }
        if (ast.value[0].value === "defmacro!") {
            if (ast.value[1].type !== "symbol") {
                throw new Error("Should be a symbol");
            }
            var evaluated_2 = EVAL(ast.value[2], env);
            if (evaluated_2.type !== "function") {
                throw new Error("should be a func");
            }
            env.set(ast.value[1].value, __assign(__assign({}, evaluated_2), { is_macro: true }));
            return { value: evaluated_2 };
        }
        if (ast.value[0].value === "let*") {
            var new_env = new env_1.Env(env);
            if (ast.value[1].type !== "list" && ast.value[1].type !== "vector") {
                throw new Error("Bindings should be a list");
            }
            var bindings = ast.value[1].value;
            if (bindings.length % 2 !== 0) {
                throw new Error("Bindings should not be odd length");
            }
            var i = 0;
            while (i < bindings.length) {
                if (bindings[i].type !== "symbol") {
                    throw new Error("Bindings should be a symbol");
                }
                else {
                    new_env.set(bindings[i].value, EVAL(bindings[i + 1], new_env));
                }
                i += 2;
            }
            ast = ast.value[2];
            env = new_env;
            return "continue";
        }
        if (ast.value[0].value === "do") {
            ast.value.slice(1, -1).map(function (val) { return EVAL(val, env); });
            ast = ast.value[ast.value.length - 1];
            return "continue";
        }
        if (ast.value[0].value === "if") {
            var cond = EVAL(ast.value[1], env);
            if (cond.value !== false && cond.type !== "nil") {
                ast = ast.value[2];
            }
            else if (ast.value.length === 4) {
                ast = ast.value[3];
            }
            else {
                ast = {
                    type: "nil"
                };
            }
            return "continue";
        }
        if (ast.value[0].value === "fn*") {
            var bindings_1 = ast.value[1];
            var content_1 = ast.value[2];
            if (bindings_1.type !== "list" && bindings_1.type !== "vector") {
                throw new Error("Function bindings should be a list");
            }
            return { value: {
                    type: "function",
                    value: {
                        params: bindings_1,
                        ast: content_1,
                        env: env,
                        fn: function () {
                            var args = [];
                            for (var _i = 0; _i < arguments.length; _i++) {
                                args[_i] = arguments[_i];
                            }
                            var new_env = new env_1.Env(env);
                            for (var i = 0; i < bindings_1.value.length; i++) {
                                if (bindings_1.value[i].type !== "symbol") {
                                    throw new Error("Function bindings should be a list");
                                }
                                // variadic arguments
                                if (bindings_1.value[i].value === "&") {
                                    new_env.set(bindings_1.value[i + 1].value, {
                                        type: "list",
                                        value: args.slice(i)
                                    });
                                    break;
                                }
                                else {
                                    new_env.set(bindings_1.value[i].value, args[i]);
                                }
                            }
                            return EVAL(content_1, new_env);
                        }
                    }
                } };
        }
        var evaluated = eval_ast(ast, env);
        if (evaluated.type !== "list") {
            throw new Error("Should be a list");
        }
        if (evaluated.value[0].type === "error") {
            return { value: evaluated.value[0] };
        }
        if (evaluated.value[0].type !== "function") {
            throw new Error("Should be a function");
        }
        var fnValue = evaluated.value[0].value;
        if (fnValue.params) {
            var args_1 = evaluated.value.slice(1);
            ast = fnValue.ast;
            var new_env = new env_1.Env(fnValue.env);
            for (var i = 0; i < fnValue.params.value.length; i++) {
                if (fnValue.params.value[i].type !== "symbol") {
                    throw new Error("Function bindings should all be args");
                }
                // variadic arguments
                if (fnValue.params.value[i].value === "&") {
                    new_env.set(fnValue.params.value[i + 1].value, {
                        type: "list",
                        value: args_1.slice(i)
                    });
                    break;
                }
                else {
                    new_env.set(fnValue.params.value[i].value, args_1[i]);
                }
            }
            env = new_env;
            return "continue";
        }
        return { value: (_a = evaluated.value[0].value).fn.apply(_a, evaluated.value.slice(1)) };
    };
    while (true) {
        var state_1 = _loop_1();
        if (typeof state_1 === "object")
            return state_1.value;
    }
};
var rep = function (input) {
    try {
        return printer_1.pr_str(EVAL(reader_1.read_str(input, true), repl_env));
    }
    catch (e) {
        if (e instanceof Error) {
            return printer_1.pr_str({
                type: "string",
                value: ".*Error.*" + e.message
            }, false);
        }
        return ".*Error.*" + printer_1.pr_str(e);
    }
};
rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\\nnil)\")))))");
rep("(def! not (fn* (a) (if a false true)))");
rep("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");
var filenameIndex = process.argv.indexOf(__filename);
var args = process.argv.slice(filenameIndex + 1);
repl_env.set("*host-language*", {
    type: "string",
    value: "ts",
});
repl_env.set("*ARGV*", {
    type: "list",
    value: []
});
if (args.length >= 1) {
    if (args.length > 1) {
        repl_env.set("*ARGV*", {
            type: "list",
            value: args.slice(1).map(function (item) { return ({
                type: "string",
                value: item
            }); })
        });
    }
    rep("(load-file \"" + args[0] + "\")");
}
else {
    while (true) {
        var line = node_readline_1.readline('user> ');
        if (line === null) {
            break;
        }
        if (line === "") {
            continue;
        }
        console.log(rep(line));
    }
}

File: impls/python/stepA_mal.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import sys
from reader import read_str, Val
from printer import pr_str
from env import Env
from core import ns


repl_env = Env()
for key in ns:
    repl_env.set(key, ns[key])
repl_env.set("eval", Val("fn", lambda a: eval(a, repl_env)))
repl_env.set("*host-language*", Val("string", "Python"))

def eval_ast(ast, env):
    if ast.type == "symbol":
        return env.get(ast.value)
    if ast.type in ("list", "vector"):
        items = map(lambda ast: eval(ast, env), ast.value)
        return Val(ast.type, items)
    if ast.type == "hashmap":
        new_val = {}
        for key, value in ast.value.items():
            new_val[key] = eval(value, env)
        return Val("hashmap", new_val)
    return ast

def is_pair(ast):
    return ast.type in ("list", "vector") and len(ast.value) > 0

def quasiquote(ast):
    if not is_pair(ast):
        return Val("list", [Val("symbol", "quote"), ast])

    if ast.value[0].value == "unquote":
        return ast.value[1]

    if is_pair(ast.value[0]) and ast.value[0].value[0].value == "splice-unquote":
        return Val("list", [
            Val("symbol", "concat"),
            ast.value[0].value[1],
            quasiquote(Val("list", ast.value[1:]))
        ])

    return Val("list", [
        Val("symbol", "cons"),
        quasiquote(ast.value[0]),
        quasiquote(Val("list", ast.value[1:]))
    ])

def is_macro_call(ast, env):
    try:
        val = env.get(ast.value[0].value)
        return val.type == 'macro'
    except:
        return False

def macroexpand(ast, env):
    while (is_macro_call(ast, env)):
        macro = env.get(ast.value[0].value)
        ast = macro.value["fn"](*ast.value[1:])
    return ast

def eval(ast, env):
    while True:
        ast = macroexpand(ast, env)
        if ast.type != "list":
            return eval_ast(ast, env)
        if len(ast.value) == 0:
            return ast

        symbol = ast.value[0].value

        if symbol == "try*":
            if (len(ast.value) <= 2):
                return eval(ast.value[1], env)
            try:
                return eval(ast.value[1], env)
            except Exception as e:
                error = Val("string", str(e))
                if isinstance(e.args[0], Val):
                    error = e.args[0]
                new_env = Env(env)

                new_env.set(ast.value[2].value[1].value, error)
                return eval(ast.value[2].value[2], new_env)

        if symbol == "macroexpand":
            return macroexpand(ast.value[1], env)

        if symbol == "quote":
            return ast.value[1]

        if symbol == "quasiquote":
            ast = quasiquote(ast.value[1])
            continue

        if symbol == "def!":
            evaluated = eval(ast.value[2], env)
            env.set(ast.value[1].value, evaluated)
            return evaluated

        if symbol == "defmacro!":
            evaluated = Val("macro", eval(ast.value[2], env).value)
            env.set(ast.value[1].value, evaluated)
            return evaluated

        if symbol == "let*":
            new_env = Env(env)
            params = ast.value[1].value

            i = 0
            while i < len(params):
                new_env.set(params[i].value, eval(params[i + 1], new_env))
                i += 2
            ast = ast.value[2]
            env = new_env
            continue

        if symbol == "do":
            for i in ast.value[1:-1]:
                eval(i, env)
            ast = ast.value[len(ast.value) -1]
            continue

        if symbol == "if":
            cond = eval(ast.value[1], env)
            if cond.type != "nil" and (cond.type != "bool" or cond.value == True):
                ast = ast.value[2]
            elif len(ast.value) > 3:
                ast = ast.value[3]
            else: 
                ast = Val("nil", [])
            continue

        if symbol == "fn*":
            def func(*exprs_t):
                new_env = Env(env)
                params = ast.value[1].value
                exprs = list(exprs_t)
                i = 0
                while i < len(params):
                    if params[i].value == "&":
                        new_env.set(params[i + 1].value, Val("list", exprs[i:]))
                        break
                    new_env.set(params[i].value, exprs[i])
                    i += 1

                return eval(ast.value[2], new_env)

            return Val("custom_fn", {
                "fn": func,
                "ast": ast.value[2],
                "params": ast.value[1].value,
                "env": env,
            })

        new_ast = eval_ast(ast, env)

        fn = new_ast.value[0]
        if fn.type == 'custom_fn':
            exprs = new_ast.value[1:]
            ast = fn.value["ast"]
            params = fn.value["params"]
            i = 0
            new_env = Env(fn.value["env"])
            while i < len(params):
                if params[i].value == "&":
                    new_env.set(params[i + 1].value, Val("list", exprs[i:]))
                    break
                new_env.set(params[i].value, exprs[i])
                i += 1
            env = new_env
            continue

        return fn.value(*new_ast.value[1:])

def rep(line):
    if len(line) == 0 or line[0] == ";":
        return
    try:
        return pr_str(eval(read_str(line), repl_env))
    except Exception as e:
        if isinstance(e.args[0], Val):
            return pr_str(Val("error", e.args[0]))
        else:
            return pr_str(Val("error", Val("string", str(e))))

rep("(def! not (fn* (a) (if a false true)))")
rep("""(def! load-file (fn* (f) (eval (read-string (str "(do " (slurp f) "\nnil)")))))""")
rep("""(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))""")

if len(sys.argv) > 1:
    content = []
    for i in sys.argv[2:]:
        content.append(Val("string", i))
    repl_env.set("*ARGV*", Val("list", content))
    rep('(load-file "' + sys.argv[1] + '")')
else:
    repl_env.set("*ARGV*", Val("list", []))
    while True:
        line = raw_input("user> ")
        res = rep(line)
        if (res != None):
            print(res)

